iree/compiler/Conversion/LinalgToSPIRV/VectorizeLinalgOps.cpp

```diff
@@ -0,0 +1,333 @@
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Conversion/HLOToLinalg/Passes.h"
+#include "iree/compiler/Dialect/IREE/IR/IREEDialect.h"
+#include "iree/compiler/Dialect/IREE/IR/IREEOps.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+/// Returns the bitwidth of a scalar or vector type.
+static Optional<unsigned> getBitWidth(Type type) {
+  if (type.isIntOrFloat()) {
+    return type.getIntOrFloatBitWidth();
+  } else if (type.isa<VectorType>()) {
+    auto vecType = type.cast<VectorType>();
+    auto elementType = vecType.getElementType();
+    return elementType.getIntOrFloatBitWidth() * vecType.getNumElements();
+  }
+  return {};
+}
+
+constexpr int kVectorizationSizeInBits = 128;
+constexpr int kVecSize = kVectorizationSizeInBits / (sizeof(float) * 8);
+
+static VectorType getVecType(OpBuilder &builder, Type t) {
+  if (t.isa<VectorType>()) return t.cast<VectorType>();
+  if (!t.isa<IntegerType, FloatType>()) return {};
+  if (t.getIntOrFloatBitWidth() != 32) return {};
+  Type newElemType = t.isa<IntegerType>() ? builder.getI32Type().cast<Type>()
+                                          : builder.getF32Type().cast<Type>();
+  return VectorType::get(kVecSize, newElemType);
+}
+
+static MemRefType getVectorizedMemRefType(OpBuilder &builder, MemRefType type) {
+  Type elemType = type.getElementType();
+  VectorType vecType = getVecType(builder, elemType);
+  if (!vecType) return {};
+  unsigned elemSize = elemType.getIntOrFloatBitWidth();
+  unsigned vecSize = kVectorizationSizeInBits / elemSize;
+  SmallVector<int64_t, 2> newShape(type.getShape().begin(),
+                                   type.getShape().end());
+  if (newShape.empty()) return {};
+  if (newShape.back() % vecSize != 0) return {};
+  newShape.back() = newShape.back() / vecSize;
+  return MemRefType::get(newShape, vecType, {}, type.getMemorySpace());
+}
+
+Value legalizeToVectorType(OpBuilder &builder, Value val) {
+  Type type = val.getType();
+  if (type.isa<VectorType>()) {
+    return val;
+  } else if (type.isIntOrFloat()) {
+    auto vecType = getVecType(builder, type);
+    if (!vecType) return {};
+    // TODO(hanchung): Add a folder on vector::BroadcastOp so we don't need to
+    // create manually.
+    if (auto cst = val.getDefiningOp<ConstantOp>()) {
```

Comment:
Another idea is to rewrite it as a slpat constant of vector size. to avoid adding  a broadcast op

Reply:
It is already doing this way. I think we still can add a broadcast op and apply canonicalize pass or create it using createOrFold, so it will still go away. This is not implemented yet.